src/Dscp/Educator/Web/Student/Queries.hs

```diff
@@ -0,0 +1,351 @@
+            return Submission{ sHash = submissionH, .. }
+  where
+    queryText clausesF = [q|
+        select    Submissions.hash, contents_hash, assignment_hash
+        from      Submissions
+        left join Assignments
+               on Assignments.hash = Submissions.assignment_hash
+        where     Submissions.student_addr = ?
+    |]
+      `filterClauses` clausesF
+
+ensureSubmissionExists
+    :: MonadStudentAPIQuery m
+    => Student -> Hash Core.Submission -> m ()
+ensureSubmissionExists = void ... getSubmission
+
+deleteSubmission
+    :: MonadStudentAPIQuery m
+    => Student
+    -> Hash Core.Submission
+    -> m ()
+deleteSubmission student submissionH = do
+    transaction $ do
+        submission <- getSubmission student submissionH
+        whenJust (sGrade submission) $ \_ ->
+            throwM DeletingGradedSubmission
+
+        execute queryText (submissionH, student)
+  where
+    queryText = [q|
+       delete
+       from     Submissions
+       where    hash = ?
+            and student_addr = ?
+    |]
+
+getBlockTxs
+    :: MonadStudentAPIQuery m
+    => Student
+    -> Word32
+    -> m [PrivateTx]
+getBlockTxs student blockIdx = do
+    query queryText (blockIdx, student)
+  where
+    queryText :: Query
+    queryText = [q|
+        select     Submissions.student_addr,
+                   Submissions.contents_hash,
+                   Assignments.course_id,
+
+                   Assignments.contents_hash,
+                   Assignments.type,
+                   Assignments.desc,
+
+                   Submissions.signature,
+                   grade,
+                   time
+
+        from       Transactions
+
+        left join  Submissions
+               on  submission_hash = Submissions.hash
+
+        left join  Assignments
+               on  assignment_hash = Assignments.hash
+
+        where      Transactions.idx = ?
+              and  student_addr = ?
+    |]
+
+getProofs
+    :: MonadStudentAPIQuery m
+    => Student
+    -> Maybe UTCTime
+    -> m [BlkProof]
+getProofs student sinceF = do
+    transaction $ do
+        blocks <- query queryText (Only sinceF)
+        forM blocks $ \(idx, tree) -> do
+            txs <- getBlockTxs student idx
+            return BlkProof
+                { bpMtreeSerialized = AsByteString tree
+                , bpTxs = txs
+                }
+  where
+    queryText :: Query
+    queryText = [q|
+        select     idx, mtree
+        from       Blocks
+        where      time >= ?
+    |]
+
+
+
+newtype FilterClause = FilterClause { unFilterClause :: String }
+    deriving (IsString)
+
+data SomeParams = forall a. ToRow a => SomeParams a
+
+instance Monoid SomeParams where
+    mempty = SomeParams ()
+    SomeParams a `mappend` SomeParams b = SomeParams (a :. b)
+
+instance ToRow SomeParams where
+    toRow (SomeParams a) = toRow a
+
+withFilterOn :: ToField a => String -> Maybe a -> ((FilterClause, SomeParams) -> b) -> b
```

Comment:
Why do you need this function to have CPS style? Seems like it only makes the code more confusing :|